File: MathCore.DSP/Filters/AnalogBasedFilter.cs
using System.Runtime.Serialization;

using static MathCore.Consts;
// ReSharper disable InconsistentNaming
// ReSharper disable ParameterHidesMember

namespace MathCore.DSP.Filters;

/// <summary>Цифровой фильтр на основе аналогового прототипа</summary>
//[KnownType(typeof(BesselFilter))]     // Наихудшая аппроксимация прямоугольной АЧХ (близка по форме к гауссовой кривой), но наилучшая форма переходной хар-ки
[KnownType(typeof(ButterworthFilter))]  // Промежуточные качества по прямоугольности АЧХ и переходной хар-ке
[KnownType(typeof(ChebyshevFilter))]    // АЧХ наиболее приближена к прямоугольной - наибольшие пульсации в переходной хор-ке
[KnownType(typeof(EllipticFilter))]     // Максимальная крутизна АЧХ
public abstract class AnalogBasedFilter : IIR
{
    //https://ru.dsplib.org/content/filter_low2low/filter_low2low.html
    /// <summary>Класс преобразований нулей и полюсов аналоговых прототипов</summary>
    public static class Transform
    {
        /// <summary>Преобразование нулей/полюсов фильтра ФНЧ-ФНЧ</summary>
        /// <param name="Z">Нули/полюса прототипа (нормированного ФНЧ)</param>
        /// <param name="wp">Новая частота среза</param>
        /// <returns>Нули/полюса нового фильтра ФНЧ</returns>
        public static IEnumerable<Complex> ToLow(IEnumerable<Complex> Z, double wp) => Z.Select(z => z / wp);

        /// <summary>Преобразование нулей/полюсов фильтра ФНЧ-ФВЧ</summary>
        /// <param name="Z">Нули/полюса прототипа (нормированного ФНЧ)</param>
        /// <param name="wp">Новая частота среза</param>
        /// <returns>Нули/полюса нового фильтра ФВЧ</returns>
        public static IEnumerable<Complex> ToHigh(IEnumerable<Complex> Z, double wp) => Z.Select(z => z * wp);

        /// <summary>Установить значения новых полюсов/нулей</summary>
        /// <param name="p">Исходное значение</param>
        /// <param name="D">Смещение</param>
        /// <param name="p1">Первое значение</param>
        /// <param name="p2">Второе значение</param>
        private static void Set(in Complex p, in Complex D, out Complex p1, out Complex p2) => (p1, p2) = (p + D, p - D);

        /// <summary>Преобразование нулей/полюсов фильтра ФНЧ-ППФ</summary>
        /// <param name="Poles">Полюса прототипа (нормированного ФНЧ)</param>
        /// <param name="Zeros">Нули прототипа (нормированного ФНЧ)</param>
        /// <param name="fpl">Нижняя частота среза ППФ</param>
        /// <param name="fph">Верхняя частота среза ППФ</param>
        /// <returns>Нули/полюса ППФ</returns>
        /// <exception cref="ArgumentException">Если число полюсов == 0</exception>
        /// <exception cref="ArgumentException">Если число нулей больше числа полюсов</exception>
        public static (Complex[] Poles, Complex[] Zeros) ToBandPass(
            Complex[] Poles,
            Complex[] Zeros,
            double fpl, double fph)
        {
            if (Poles is null) throw new ArgumentNullException(nameof(Poles));
            if (Zeros is null) throw new ArgumentNullException(nameof(Zeros));
            if (Poles.Length == 0) throw new ArgumentException("Размер вектора полюсов должен быть больше 0", nameof(Poles));
            if (Zeros.Length > Poles.Length) throw new ArgumentException("Число нулей не должна быть больше числа полюсов", nameof(Zeros));

            var (wpl, wph) = (pi2 * fpl, pi2 * fph);
            var (dw05, wc2) = ((wph - wph) / 2, wpl * wph);

            // На каждый исходный полюс формируется пара новых полюсов
            // Число нулей равно удвоенному числу исходных нулей
            //      + ноль в 0 кратности, равной разности числа полюсов и нулей
            var poles = new Complex[Poles.Length * 2];
            var zeros = new Complex[Zeros.Length * 2 + (Poles.Length - Zeros.Length)];


            for (var i = 0; i < Poles.Length; i++)
            {
                var pdw = dw05 * Poles[i];
                Set(pdw, Complex.Sqrt(pdw.Power - wc2),
                    out poles[2 * i],
                    out poles[2 * i + 1]);
            }

            for (var i = 0; i < Zeros.Length; i++)
            {
                var pdw = dw05 * Zeros[i];
                Set(pdw, Complex.Sqrt(pdw.Power - wc2),
                    out zeros[2 * i],
                    out zeros[2 * i + 1]);
            }

            return (poles, zeros);
        }

        /// <summary>Преобразование нулей/полюсов фильтра ФНЧ-ПЗФ</summary>
        /// <param name="Poles">Полюса прототипа (нормированного ФНЧ)</param>
        /// <param name="Zeros">Нули прототипа (нормированного ФНЧ)</param>
        /// <param name="fpl">Нижняя частота среза ПЗФ</param>
        /// <param name="fph">Верхняя частота среза ПЗФ</param>
        /// <returns>Нули/полюса ППФ</returns>
        /// <exception cref="ArgumentException">Если число полюсов == 0</exception>
        /// <exception cref="ArgumentException">Если число нулей больше числа полюсов</exception>
        public static (Complex[] Poles, Complex[] Zeros) ToBandStop(
            Complex[] Poles,
            Complex[] Zeros,
            double fpl, double fph)
        {
            if (Poles is null) throw new ArgumentNullException(nameof(Poles));
            if (Zeros is null) throw new ArgumentNullException(nameof(Zeros));
            var count_p = Poles.Length;
            var count_0 = Zeros.Length;
            if (count_p == 0) throw new ArgumentException("Размер вектора полюсов должен быть больше 0", nameof(Poles));
            if (count_0 > count_p) throw new ArgumentException("Число нулей не должна быть больше числа полюсов", nameof(Zeros));

            var (wpl, wph) = (pi2 * fpl, pi2 * fph);
            var (dw05, wc2) = ((wph - wph) / 2, wpl * wph);

            // На каждый исходный полюс формируется пара новых полюсов
            // Число нулей равно удвоенному числу исходных нулей
            //      + ноль в 0 кратности, равной разности числа полюсов и нулей
            var poles = new Complex[count_p * 2];
            var zeros = new Complex[poles.Length];

            //static void Set(in Complex p, in Complex D, out Complex p1, out Complex p2)
            //{
            //    p1 = p + D;
            //    p2 = p - D;
            //}

            for (var i = 0; i < count_p; i++)
            {
                var pdw = dw05 / Poles[i];
                Set(pdw, Complex.Sqrt(pdw.Power - wc2),
                    out poles[2 * i],
                    out poles[2 * i + 1]);
            }

            for (var i = 0; i < count_0; i++)
            {
                var pdw = dw05 / Zeros[i];
                Set(pdw, Complex.Sqrt(pdw.Power - wc2),
                    out zeros[2 * i],
                    out zeros[2 * i + 1]);
            }

            var wc = wc2.Sqrt();
            for (var i = 0; i < count_p - count_0; i++)
            {
                zeros[2 * (count_0 + i)] = new(0, wc);
                zeros[2 * (count_0 + i) + 1] = new(0, -wc);
            }

            return (poles, zeros);
        }
    }

    /// <summary>Спецификация фильтра</summary>
    public readonly ref struct Specification
    {
        /// <summary>Период дискретизации</summary>
        public double dt { get; }

        /// <summary>Частота дискретизации</summary>
        public double fd => 1 / dt;

        /// <summary>Граничная частота полосы пропускания</summary>
        public double fp { get; }
        /// <summary>Граничная частота полосы заграждения</summary>
        public double fs { get; }

        /// <summary>Граничная циклическая частота полосы пропускания</summary>
        public double wp { get; }
        /// <summary>Граничная циклическая частота полосы заграждения</summary>
        public double ws { get; }

        /// <summary>Граничная частота полосы пропускания цифрового фильтра</summary>
        public double Fp { get; }
        /// <summary>Граничная частота полосы заграждения цифрового фильтра</summary>
        public double Fs { get; }

        /// <summary>Граничная циклическая частота полосы пропускания цифрового фильтра</summary>
        public double Wp { get; }
        /// <summary>Граничная циклическая частота полосы заграждения цифрового фильтра</summary>
        public double Ws { get; }

        /// <summary>Коэффициент передачи в полосе пропускания</summary>
        public double Gp { get; }
        /// <summary>Коэффициент передачи в полосе подавления</summary>
        public double Gs { get; }

        /// <summary>Коэффициент подавления в полосе пропускания (дБ)</summary>
        public double Rp { get; }
        /// <summary>Коэффициент подавления в полосе заграждения (дБ)</summary>
        public double Rs { get; }

        /// <summary>Неоднородность АЧХ в полосе пропускания</summary>
        public double EpsP { get; }
        /// <summary>Неоднородность АЧХ в полосе заграждения</summary>
        public double EpsS { get; }

        /// <summary>Отношение неоднородностей коэффициентов передачи заграждения и пропускания</summary>
        public double kEps => EpsS / EpsP;
        /// <summary>Отношение частотных полос заграждения и пропускания</summary>
        public double kw => fs / fp;
        /// <summary>Отношение частотных полос заграждения и пропускания цифрового фильтра</summary>
        public double kW => Ws / Wp;

        /// <summary>Инициализация новой спецификации фильтра</summary>
        /// <param name="dt">Период дискретизации</param>
        /// <param name="fp">Граничная частота полосы пропускания</param>
        /// <param name="fs">Граничная частота полосы заграждения</param>
        /// <param name="Gp">Коэффициент передачи в полосе пропускания</param>
        /// <param name="Gs">Коэффициент передачи в полосе заграждения</param>
        /// <exception cref="InvalidOperationException">Если частота пропускания больше, либо равна частоте подавления</exception>
        /// <exception cref="InvalidOperationException">Если частота пропускания больше, либо равна половине частоты дискретизации</exception>
        public Specification(double dt, double fp, double fs, double Gp, double Gs)
        {
            if (!(fp < 1 / (2 * dt))) throw new InvalidOperationException("Частота пропускания должен быть меньше половины полосы дискретизации");
            if (fp < 0) throw new ArgumentOutOfRangeException(nameof(fp), fp, "Частота среза не может быть меньше 0");
            if (fs < 0) throw new ArgumentOutOfRangeException(nameof(fs), fs, "Частота полосы подавления не может быть меньше 0");
            if (Gp < 0) throw new ArgumentOutOfRangeException(nameof(Gp), Gp, "Значение уровня АЧХ в полосе пропускания не может быть отрицательной величиной");
            if (Gs < 0) throw new ArgumentOutOfRangeException(nameof(Gs), Gs, "Значение уровня АЧХ в полосе заграждения не может быть отрицательной величиной");
            if (Gp > 1) throw new ArgumentOutOfRangeException(nameof(Gp), Gp, "Значение уровня АЧХ в полосе пропускания не может быть больше 1");
            if (Gs > 1) throw new ArgumentOutOfRangeException(nameof(Gs), Gs, "Значение уровня АЧХ в полосе заграждения не может быть больше 1");
            if (Gp <= Gs) throw new InvalidOperationException($"Уровень АЧХ в полосе пропускания Gp должен быть больше уровня в полосе заграждения Gs\r\n  Gp={Gp}\r\n  Gs={Gs}");

            this.dt = dt;
            this.fp = fp;
            this.fs = fs;
            this.Gp = Gp;
            this.Gs = Gs;

            Rp = -Gp.In_dB();
            Rs = -Gs.In_dB();

            EpsP = (1 / (Gp * Gp) - 1).Sqrt();
            //EpsP = (10d.Pow(Rp / 10) - 1).Sqrt();
            EpsS = (10d.Pow(Rs / 10) - 1).Sqrt();

            //var tEpsP = (1 / Gp.Pow2() - 1).Sqrt();
            //var tEpsS = (1 / Gs.Pow2() - 1).Sqrt();

            //var tEpsP = (1 - Gp*Gp).Sqrt() / (Gp*Gp);
            //var tEpsS = (1 - Gs*Gs).Sqrt() / (Gs*Gs);

            Fp = ToDigitalFrequency(fp, dt);  // Частота пропускания аналогового прототипа
            Fs = ToDigitalFrequency(fs, dt);  // Частота подавления аналогового пропита

            (wp, ws) = (pi2 * fp, pi2 * fs);
            (Wp, Ws) = (pi2 * Fp, pi2 * Fs);
        }

        /// <summary>Деконструктор спецификации фильтра</summary>
        /// <param name="dt">Период дискретизации</param>
        /// <param name="fp">Граничная частота полосы пропускания</param>
        /// <param name="fs">Граничная частота полосы заграждения</param>
        /// <param name="Gp">Коэффициент передачи в полосе пропускания</param>
        /// <param name="Gs">Коэффициент передачи в полосе заграждения</param>
        public void Deconstruct(out double dt, out double fp, out double fs, out double Gp, out double Gs) =>
            (dt, fp, fs, Gp, Gs) = (this.dt, this.fp, this.fs, this.Gp, this.Gs);
    }

    /// <summary>Создать спецификацию фильтра</summary>
    /// <param name="dt">Период дискретизации</param>
    /// <param name="fp">Граничная частота полосы пропускания</param>
    /// <param name="fs">Граничная частота полосы заграждения</param>
    /// <param name="Gp">Коэффициент передачи в полосе пропускания</param>
    /// <param name="Gs">Коэффициент передачи в полосе заграждения</param>
    /// <returns>Спецификация фильтра</returns>
    public static Specification GetSpecification(double dt, double fp, double fs, double Gp, double Gs) => new(dt, fp, fs, Gp, Gs);

    /// <summary>Период дискретизации</summary>
    public double dt { get; }

    /// <summary>Частота дискретизации</summary>
    public double fd => 1 / dt;

    /// <summary>Граничная частота полосы пропускания</summary>
    public double fp { get; }

    /// <summary>Граничная частота полосы подавления</summary>
    public double fs { get; }

    /// <summary>Затухание в полосе пропускания</summary>
    public double Gp { get; }

    /// <summary>Затухание в полосе подавления</summary>
    public double Gs { get; }

    /// <summary>Спецификация фильтра</summary>
    public Specification Spec => new(dt, fp, fs, Gp, Gs);

    /// <summary>Инициализация параметров цифрового фильтра на базе аналогового прототипа</summary>
    /// <param name="B">Коэффициенты полинома числителя</param>
    /// <param name="A">Коэффициенты полинома знаменателя</param>
    /// <param name="Spec">Спецификация фильтра</param>
    protected AnalogBasedFilter(double[] B, double[] A, Specification Spec) : base(B, A) => (dt, fp, fs, Gp, Gs) = Spec;

    /// <inheritdoc/>
    public override Complex FrequencyResponse(double f) => base.FrequencyResponse(f / fd);

    /// <inheritdoc/>
    public override Complex FrequencyResponse(double f, double dt) => base.FrequencyResponse(f * dt);

    /// <summary>Метод преобразования нулей и полюсов нормированного ФНЧ в нули и полюса ППФ</summary>
    /// <param name="Normed">Нормированные нули и полюса ФНЧ</param>
    /// <param name="fmin">Нижняя частота среза</param>
    /// <param name="fmax">Верхняя частота среза</param>
    /// <returns>Нули и полюса ППФ</returns>
    public static IEnumerable<Complex> TransformToBandPass(IEnumerable<Complex> Normed, double fmin, double fmax) =>
        TransformToBandPassW(Normed, pi2 * fmin, pi2 * fmax);

    /// <summary>Метод преобразования нулей и полюсов нормированного ФНЧ в нули и полюса ППФ</summary>
    /// <param name="Normed">Нормированные нули и полюса ФНЧ</param>
    /// <param name="w_min">Нижняя частота среза</param>
    /// <param name="w_max">Верхняя частота среза</param>
    /// <returns>Нули и полюса ППФ</returns>
    public static IEnumerable<Complex> TransformToBandPassW(IEnumerable<Complex> Normed, double w_min, double w_max)
    {
        var (dw05, wc2) = ((w_max - w_min) / 2, w_min * w_max);

        foreach (var p in Normed)
        {
            var pdw = dw05 * p;
            var sqrt = Complex.Sqrt(pdw.Pow2() - wc2);
            yield return pdw + sqrt;
            yield return pdw - sqrt;
        }
    }

    /// <summary>Метод преобразования нулей и полюсов нормированного ФНЧ в нули и полюса ПЗФ</summary>
    /// <param name="Normed">Нормированные нули и полюса ФНЧ</param>
    /// <param name="fmin">Нижняя частота среза</param>
    /// <param name="fmax">Верхняя частота среза</param>
    /// <returns>Нули и полюса ПЗФ</returns>
    public static IEnumerable<Complex> TransformToBandStop(IEnumerable<Complex> Normed, double fmin, double fmax) =>
        TransformToBandStopW(Normed, pi2 * fmin, pi2 * fmax);

    /// <summary>Метод преобразования нулей и полюсов нормированного ФНЧ в нули и полюса ПЗФ</summary>
    /// <param name="Normed">Нормированные нули и полюса ФНЧ</param>
    /// <param name="w_min">Нижняя частота среза</param>
    /// <param name="w_max">Верхняя частота среза</param>
    /// <returns>Нули и полюса ПЗФ</returns>
    public static IEnumerable<Complex> TransformToBandStopW(IEnumerable<Complex> Normed, double w_min, double w_max)
    {
        var (dw05, wc2) = ((w_max - w_min) / 2, w_min * w_max);

        foreach (var p in Normed)
        {
            var pdw = dw05 / p;
            var sqrt = Complex.Sqrt(pdw.Pow2() - wc2);
            yield return pdw + sqrt;
            yield return pdw - sqrt;
        }
    }

    /// <summary>Преобразование нулей/полюсов нормированного ФНЧ в нули/полюса ФНЧ с другой частотой среза</summary>
    /// <param name="Normed">Нормированные нули/полюса ФНЧ</param>
    /// <param name="fp">Новая частота среза</param>
    /// <returns>Нули/полюса нового ФНЧ</returns>
    public static IEnumerable<Complex> TransformToLowPass(IEnumerable<Complex> Normed, double fp) =>
        TransformToLowPassW(Normed, pi2 * fp);

    /// <summary>Преобразование нулей/полюсов нормированного ФНЧ в нули/полюса ФНЧ с другой частотой среза (в циклических частотах)</summary>
    /// <param name="Normed">Нормированные нули/полюса ФНЧ</param>
    /// <param name="wp">Новая циклическая частота среза</param>
    /// <returns>Нули/полюса нового ФНЧ</returns>
    public static IEnumerable<Complex> TransformToLowPassW(IEnumerable<Complex> Normed, double wp)
    {
        foreach (var p in Normed)
            yield return wp * p;
    }

    /// <summary>Преобразование нулей/полюсов нормированного ФНЧ в нули/полюса ФВЧ с другой частотой среза</summary>
    /// <param name="Normed">Нормированные нули/полюса ФНЧ</param>
    /// <param name="fp">Новая частота среза</param>
    /// <returns>Нули/полюса нового ФВЧ</returns>
    public static IEnumerable<Complex> TransformToHighPass(IEnumerable<Complex> Normed, double fp) =>
        TransformToHighPassW(Normed, pi2 * fp);

    /// <summary>Преобразование нулей/полюсов нормированного ФНЧ в нули/полюса ФВЧ с другой частотой среза (в циклических частотах)</summary>
    /// <param name="Normed">Нормированные нули/полюса ФНЧ</param>
    /// <param name="wp">Новая циклическая частота среза</param>
    /// <returns>Нули/полюса нового ФВЧ</returns>
    public static IEnumerable<Complex> TransformToHighPassW(IEnumerable<Complex> Normed, double wp)
    {
        foreach (var p in Normed)
            yield return wp / p;
    }
}

File: MathCore.DSP/Filters/Python/filters/analog_based_filter.py
# Базовый класс фильтра на основе аналогового прототипа (порт C# AnalogBasedFilter)
import numpy as np
from .iir import IIR
from .digital_filter import DigitalFilter

pi2 = 2.0 * np.pi  # 2π — часто используемая константа


class Specification:
    """
    Спецификация цифрового фильтра на основе аналогового прототипа.

    Хранит все параметры фильтра и вычисляет производные величины:
    частоты в аналоговой и цифровой областях, коэффициенты неравномерности АЧХ.
    """

    def __init__(self, dt: float, fp: float, fs: float, Gp: float, Gs: float):
        """
        Параметры
        ---------
        dt : период дискретизации (с)
        fp : граничная частота полосы пропускания (Гц)
        fs : граничная частота полосы заграждения (Гц)
        Gp : коэффициент передачи на границе пропускания (0 < Gs < Gp ≤ 1)
        Gs : коэффициент передачи на границе заграждения
        """
        if not (fp < 1.0 / (2.0 * dt)):
            raise ValueError("Частота пропускания fp должна быть меньше половины частоты дискретизации")
        if fp < 0:
            raise ValueError(f"Частота среза fp не может быть отрицательной: fp={fp}")
        if not (0 < Gs < Gp <= 1):
            raise ValueError(f"Требуется 0 < Gs < Gp ≤ 1, получено Gp={Gp}, Gs={Gs}")

        self.dt = dt
        self.fp = fp
        self.fs = fs
        self.Gp = Gp
        self.Gs = Gs

        # Затухания в дБ
        self.Rp = -20.0 * np.log10(Gp)   # потери в полосе пропускания
        self.Rs = -20.0 * np.log10(Gs)   # затухание в полосе заграждения

        # Неоднородности АЧХ (параметры аналогового прототипа)
        self.EpsP = np.sqrt(1.0 / (Gp * Gp) - 1.0)
        self.EpsS = np.sqrt(10.0 ** (self.Rs / 10.0) - 1.0)

        # Пред-искажённые частоты аналогового прототипа
        self.Fp = DigitalFilter.to_digital_frequency(fp, dt)
        self.Fs = DigitalFilter.to_digital_frequency(fs, dt)

        # Циклические частоты оригинального (аналогового) фильтра
        self.wp = pi2 * fp
        self.ws = pi2 * fs

        # Циклические частоты аналогового прототипа (пред-искажённые)
        self.Wp = pi2 * self.Fp
        self.Ws = pi2 * self.Fs

    @property
    def fd(self) -> float:
        """Частота дискретизации (Гц)."""
        return 1.0 / self.dt

    @property
    def kEps(self) -> float:
        """Отношение неоднородностей EpsS / EpsP."""
        return self.EpsS / self.EpsP

    @property
    def kw(self) -> float:
        """Отношение аналоговых частот fs / fp."""
        return self.fs / self.fp

    @property
    def kW(self) -> float:
        """Отношение цифровых частот Ws / Wp."""
        return self.Ws / self.Wp

    def __repr__(self) -> str:
        return (f"Specification(fd={self.fd:.1f} Гц, fp={self.fp:.1f} Гц, "
                f"fs={self.fs:.1f} Гц, Gp={self.Gp:.4f}, Gs={self.Gs:.4f})")


class AnalogBasedFilter(IIR):
    """
    Цифровой фильтр на основе аналогового прототипа.

    Реализует билинейное z-преобразование и трансформации нулей/полюсов
    нормированного ФНЧ в различные типы фильтров (ФНЧ, ФВЧ, ППФ, ПЗФ).
    """

    def __init__(self, B: np.ndarray, A: np.ndarray, spec: Specification):
        super().__init__(B, A)
        self.dt = spec.dt
        self.fp = spec.fp
        self.fs = spec.fs
        self.Gp = spec.Gp
        self.Gs = spec.Gs

    @property
    def fd(self) -> float:
        """Частота дискретизации (Гц)."""
        return 1.0 / self.dt

    @property
    def spec(self) -> Specification:
        """Спецификация фильтра."""
        return Specification(self.dt, self.fp, self.fs, self.Gp, self.Gs)

    # ── Статические методы получения спецификации ─────────────────────────────

    @staticmethod
    def get_specification(dt: float, fp: float, fs: float,
                          Gp: float = 0.891250938, Gs: float = 0.01) -> Specification:
        """Создаёт спецификацию фильтра."""
        return Specification(dt, fp, fs, Gp, Gs)

    # ── Трансформации нулей/полюсов ───────────────────────────────────────────

    @staticmethod
    def transform_to_low_pass_w(normed, wp: float) -> list:
        """ФНЧ-прототип → ФНЧ с частотой wp (в циклических частотах): p → p·wp."""
        return [p * wp for p in normed]

    @staticmethod
    def transform_to_high_pass_w(normed, wp: float) -> list:
        """ФНЧ-прототип → ФВЧ с частотой wp (в циклических частотах): p → wp/p."""
        return [wp / p for p in normed]

    @staticmethod
    def transform_to_band_pass_w(normed, w_min: float, w_max: float) -> list:
        """
        ФНЧ-прототип → ППФ (в циклических частотах).

        Каждый полюс/нуль p → пара (p·dW/2 ± sqrt((p·dW/2)² − Wc²)).
        """
        dw05 = (w_max - w_min) / 2.0
        wc2 = w_min * w_max
        result = []
        for p in normed:
            pdw = dw05 * p
            sqrt_val = np.sqrt(complex(pdw ** 2 - wc2))
            result.append(pdw + sqrt_val)
            result.append(pdw - sqrt_val)
        return result

    @staticmethod
    def transform_to_band_stop_w(normed, w_min: float, w_max: float) -> list:
        """
        ФНЧ-прототип → ПЗФ (в циклических частотах).

        Каждый полюс/нуль p → пара (dW/(2p) ± sqrt((dW/(2p))² − Wc²)).
        """
        dw05 = (w_max - w_min) / 2.0
        wc2 = w_min * w_max
        result = []
        for p in normed:
            pdw = dw05 / p
            sqrt_val = np.sqrt(complex(pdw ** 2 - wc2))
            result.append(pdw + sqrt_val)
            result.append(pdw - sqrt_val)
        return result

    @staticmethod
    def transform_to_low_pass(normed, fp: float) -> list:
        """ФНЧ-прототип → ФНЧ (в обычных частотах Гц)."""
        return AnalogBasedFilter.transform_to_low_pass_w(normed, pi2 * fp)

    @staticmethod
    def transform_to_high_pass(normed, fp: float) -> list:
        """ФНЧ-прототип → ФВЧ (в обычных частотах Гц)."""
        return AnalogBasedFilter.transform_to_high_pass_w(normed, pi2 * fp)

    @staticmethod
    def transform_to_band_pass(normed, fmin: float, fmax: float) -> list:
        """ФНЧ-прототип → ППФ (в обычных частотах Гц)."""
        return AnalogBasedFilter.transform_to_band_pass_w(normed, pi2 * fmin, pi2 * fmax)

    @staticmethod
    def transform_to_band_stop(normed, fmin: float, fmax: float) -> list:
        """ФНЧ-прототип → ПЗФ (в обычных частотах Гц)."""
        return AnalogBasedFilter.transform_to_band_stop_w(normed, pi2 * fmin, pi2 * fmax)

    # ── Вспомогательные методы ────────────────────────────────────────────────

    @staticmethod
    def get_binomial_coefficients(N: int) -> np.ndarray:
        """Биномиальные коэффициенты C(N,0), C(N,1), …, C(N,N) без целочисленного переполнения."""
        if N < 0:
            raise ValueError(f"Порядок бинома N не может быть отрицательным: N={N}")
        coeffs = np.zeros(N + 1, dtype=float)
        coeffs[0] = 1.0
        for k in range(1, N + 1):
            coeffs[k] = coeffs[k - 1] * (N - (k - 1)) / k
        return coeffs

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}(fd={self.fd:.1f} Гц, fp={self.fp:.1f} Гц, fs={self.fs:.1f} Гц)"
